{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2c592c87",
   "metadata": {},
   "source": [
    "# Assigment 2 - Python dictionaries and file I/O\n",
    "\n",
    "### Overview\n",
    "------------\n",
    "\n",
    "The GTEx gene expression dataset consists of samples taken from many human subjects across a variety of tissues. From this dataset we have identified a set of highly-expressed tissue-specific genes. In this exercise, you will be pulling the expressing data for those genes but only in their primary tissue of expression and investigating how much variability exists across the sample population.\n",
    "\n",
    "Biological Learning Objectives\n",
    "\n",
    "- Use metadata to associate data with subject or sample traits\n",
    "- Use descriptive statistics to make inferences about gene expression heterogeneity\n",
    "\n",
    "Computational Learning Objectives\n",
    "\n",
    "- Create and modify a dicitionary\n",
    "- Use keys to retrieve values stored in a dictionary\n",
    "- Step through each key and value in a dictionary inside a `for` loop\n",
    "- Open and read a text file\n",
    "- Write to a text file\n",
    "\n",
    "### Instructions\n",
    "----------------\n",
    "\n",
    "- Save a copy of this notebook as `~/qbXX-answers/python_dictionaries.ipynb`.\n",
    "- Fill in answers in the available code/markdown cells below.\n",
    "- Remember to comment your code to help yourself and us know what each part is intended to do."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ba1d85fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use this dictionary for identifying genes of interest and their corresponding tissue of interest\n",
    "\n",
    "gene_tissue = {\n",
    "    'ENSG00000042832.11': 'Thyroid',\n",
    "    'ENSG00000091704.9': 'Pancreas',\n",
    "    'ENSG00000118245.2': 'Testis',\n",
    "    'ENSG00000122304.10': 'Testis',\n",
    "    'ENSG00000122852.14': 'Lung',\n",
    "    'ENSG00000132693.12': 'Liver',\n",
    "    'ENSG00000134812.7': 'Stomach',\n",
    "    'ENSG00000135346.8': 'Pituitary',\n",
    "    'ENSG00000137392.9': 'Pancreas',\n",
    "    'ENSG00000142515.14': 'Prostate',\n",
    "    'ENSG00000142615.7': 'Pancreas',\n",
    "    'ENSG00000142789.19': 'Pancreas',\n",
    "    'ENSG00000158874.11': 'Liver',\n",
    "    'ENSG00000162438.11': 'Pancreas',\n",
    "    'ENSG00000164816.7': 'Small Intestine - Terminal Ileum',\n",
    "    'ENSG00000164822.4': 'Small Intestine - Terminal Ileum',\n",
    "    'ENSG00000168925.10': 'Pancreas',\n",
    "    'ENSG00000168928.12': 'Pancreas',\n",
    "    'ENSG00000170890.13': 'Pancreas',\n",
    "    'ENSG00000171195.10': 'Minor Salivary Gland',\n",
    "    'ENSG00000172179.11': 'Pituitary',\n",
    "    'ENSG00000175535.6': 'Pancreas',\n",
    "    'ENSG00000175646.3': 'Testis',\n",
    "    'ENSG00000182333.14': 'Stomach',\n",
    "    'ENSG00000187021.14': 'Pancreas',\n",
    "    'ENSG00000203784.2': 'Testis',\n",
    "    'ENSG00000204983.13': 'Pancreas',\n",
    "    'ENSG00000219073.7': 'Pancreas',\n",
    "    'ENSG00000229859.9': 'Stomach',\n",
    "    'ENSG00000240338.5': 'Pancreas',\n",
    "    'ENSG00000254647.6': 'Pancreas',\n",
    "    'ENSG00000256713.7': 'Stomach',\n",
    "    'ENSG00000259384.6': 'Pituitary',\n",
    "    }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b9449c2",
   "metadata": {},
   "source": [
    "1. Load sample metadata from \"~/Data/GTEx/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt\" and create a dictionary of sample metadata associating each sample ID with the tissue it was sampled from\n",
    "\n",
    "    - This file is tab-separated but some values have spaces in their text so make sure to specify `\\t` in the `.split()` method\n",
    "    - There are several columns in the metadata file, but you only need to be concerned with `SAMPID` and `SMTSD`, the sample ID and sample tissue, respectively\n",
    "    - Your dictionary should use the `SAMPID` as the key and `SMTSD` as the value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8f7622d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_fname = \"/Users/cmdb/Data/GTEx/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt\"\n",
    "\n",
    "# Create a dictionary to hold the metadata\n",
    "data_dict = {}\n",
    "\n",
    "# Step through each line of the metadata file\n",
    "for line in open(metadata_fname):\n",
    "    \n",
    "    # Remove the newline character from the line you just read in and split it by the tab character\n",
    "    line = line.strip(\"\\n\")\n",
    "    line_list = line.split(\"\\t\")\n",
    "    \n",
    "    # Add the data to your metadata dictionary, using the \"SAMPID\" column value as the dictionary key and the \"SMTSD\" column value as the dictionary value\n",
    "    if line_list[0].startswith(\"SAMP\"):\n",
    "        continue\n",
    "    data_dict[line_list[0]] = line_list[6]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1821cf32",
   "metadata": {},
   "source": [
    "2. Load and process the headeer from the GTEx gene expression file \"~/Data/GTEx/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct.gz\"\n",
    "\n",
    "    - You have code that will open the file and read in the header line\n",
    "    - Create a list of tissues corresponding to the column names in the header, one tissue per column name (don't worry about which the column is a sample ID or not since column names that aren't sample IDs will just get a \"missing\" label)\n",
    "    - For each value in the header, see if it appears in the dictionary of sample ID/tissues that you created in step 1\n",
    "    - If the value is in your dictionary, add the corresponding tissue name (the value in dictionary for that sample ID key) to your tissue list\n",
    "    - If the value is not in your dictionary, add \"missing\" to the tissue list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "544a432c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Because this file is a gzip-compressed file, you will be using the `gzip` library that comes with python\n",
    "import gzip\n",
    "\n",
    "expression_fname = \"/Users/cmdb/Data/GTEx/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct.gz\"\n",
    "\n",
    "# Open the file using the gzip library\n",
    "fs = gzip.open(expression_fname)\n",
    "\n",
    "# The file starts with two lines of information before the header, so these are skipped using 2 calls of `.readline()` without keeping the returned data\n",
    "_ = fs.readline()\n",
    "_ = fs.readline()\n",
    "\n",
    "# Gzipped text files are read in as byte strings, not regular strings, so you will see that `.decode()` is included after `.readline()` to convert the input into a string\n",
    "header = fs.readline().decode().rstrip().split(\"\\t\")\n",
    "\n",
    "# Make sure to close the file when you are done reading it\n",
    "fs.close()\n",
    "\n",
    "# Create a list to hold the tissue names\n",
    "tissues = []\n",
    "\n",
    "i = 0\n",
    "# Step through each value in the header\n",
    "for i in range(len(header)):\n",
    "\n",
    "    # Check if the value has an entry in the gene_tissue dictionary, adding either the tissue corresponding to that entry or \"missing\" to your tissues list\n",
    "    if header[i] in data_dict.keys():\n",
    "        tissues.append(data_dict[header[i]])\n",
    "    else:\n",
    "        tissues.append(\"missing\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eed39f12",
   "metadata": {},
   "source": [
    "3. Create a list of expression values for each gene in the `gene_tissue` dictionary (given at the start of the exercise), keeping only expression values from the tissue of interest for that gene\n",
    "\n",
    "    - Since the file is already open, you can keep reading lines from it in, picking up from after the last line read\n",
    "    - You will be log-transforming the expression data using the `log2` function from the built-in library `math`\n",
    "    - To deal with zeros in expression data, add one before log-transforming expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5b44a45b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import the log2 function from the math library to transform the expression counts\n",
    "from math import log2\n",
    "\n",
    "# Create a dictionary to hold each gene's expression data\n",
    "gene_expression = {}\n",
    "\n",
    "# Because this is a large file, for testing purposes you can use only a small portion of the file to test your code\n",
    "line_counter = 0\n",
    "\n",
    "i = 0\n",
    "# Step through each line of the file\n",
    "for line in gzip.open(expression_fname):\n",
    "\n",
    "    ############# REMOVE THIS AND RERUN ONCE YOUR CODE IS WORKING ####################\n",
    "    # For debugging, use only the first 500 lines of the file.\n",
    "    # line_counter += 1\n",
    "    # if line_counter == 500:\n",
    "    #     break\n",
    "    ############# REMOVE THIS AND RERUN ONCE YOUR CODE IS WORKING ####################\n",
    "    \n",
    "    # Convert the line into a string from a byte string, remove the newline character at the end, and split it by tab characters\n",
    "    fields = line.decode().rstrip().split(\"\\t\")\n",
    "    \n",
    "    # Check if this line contains data for a gene we are interested in. If not, skip it\n",
    "    if fields[0] not in gene_tissue.keys():\n",
    "        continue\n",
    "        \n",
    "    # Determine which tissue we care about for this gene\n",
    "    fields_tissue = gene_tissue[fields[0]]\n",
    "\n",
    "    # Create a list in the gene expression dictionary to hold the gene's data\n",
    "    gene_expression[fields[0]] = []\n",
    "    \n",
    "    # Look at each column position, one by one\n",
    "    for i in range(len(tissues)):\n",
    "\n",
    "    # For each column, see if it is from the tissue we care about (using the tissues list). If so, add that field to our gene expression list, first transforming it with log2(1 + float(field[i])) \n",
    "        if tissues[i] == fields_tissue:\n",
    "            gene_expression[fields[0]].append(log2(1 + float(fields[i])))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4676fc06",
   "metadata": {},
   "source": [
    "4. For each gene, calculate the mean, standard deviation, and relative standard deviation and write them to a results file\n",
    "\n",
    "    - You will be using the `mean` and `stdev` functions from the built-in `statistics` library. Both functions accept a list as their input\n",
    "    - Format your results for each gene into a string and write that string to your results file\n",
    "    - If you want to organize your results better (completely optional!!), you can save your results in a list and sorting them by tissue type before writing them to a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6b89b161",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the functions to caculcate the mean and standard deviation from a list\n",
    "from statistics import stdev, mean\n",
    "\n",
    "# Open a file to write the results to\n",
    "result = open(\"/Users/cmdb/qbb2025/bootcamp/day2_morning/result.txt\", \"w\")\n",
    "\n",
    "num = 0\n",
    "list_result = []\n",
    "# Step through each gene and its expression value list\n",
    "for gene in gene_expression.keys():\n",
    "\n",
    "    # Calculate the mean and standard deviation of expression\n",
    "    mean_res = mean(gene_expression[gene])\n",
    "    sd_res = stdev(gene_expression[gene])\n",
    "\n",
    "    # Calculate the relative standard deviation (coefficient of variation, std / mean)\n",
    "    cov_res = sd_res / mean_res\n",
    "\n",
    "    # Write the tissue namne, gene name, mean, standard deviation, and relative standard deviation to your results file\n",
    "    list_result.append([gene_tissue[gene], gene, mean_res, sd_res, cov_res])\n",
    "\n",
    "list_result = sorted(list_result)\n",
    "\n",
    "for item in list_result:\n",
    "    item[2] = str(item[2])\n",
    "    item[3] = str(item[3])\n",
    "    item[4] = str(item[4])\n",
    "    print(\"\\t\".join(item), file = result)\n",
    "\n",
    "# Close your results file\n",
    "result.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "185fe1ce",
   "metadata": {},
   "source": [
    "5. Looking at your results, answer the following questions\n",
    "\n",
    "    - Which tissues have low relative variability (rv < 0.2), medium relative variability (0.2 < rv < 0.4), or high relative variability (rv > 0.4)?\n",
    "    - Do each of the genes for a given tissue show the same level of relative variability?\n",
    "    - Propose an explanation for differences in relative variability you see between tissues based on what you know about their functions.\n",
    "    - Were any of the results surpising to you?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "570f5d34",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tissues with low relative variability: Thyroid, Liver, Pancreas, Testis, Pituitary\n",
      "Tissues with medium relative variability: Small Intestine - Terminal Ileum, Lung, Liver, Minor Salivary Gland, Prostate\n",
      "Tissues with high relative variability: Stomach\n"
     ]
    }
   ],
   "source": [
    "# Set empty lists to collect tissue names\n",
    "low_tissue = []\n",
    "medium_tissue = []\n",
    "high_tissue = []\n",
    "\n",
    "# Put tissues into three categories: low rv, medium rv, and high rv\n",
    "for item in list_result:\n",
    "    name = item[0]\n",
    "    cov = float(item[4])\n",
    "    if cov < 0.2:\n",
    "        low_tissue.append(name)\n",
    "    elif 0.2 < cov < 0.4:\n",
    "        medium_tissue.append(name)\n",
    "    elif cov > 0.4:\n",
    "        high_tissue.append(name)\n",
    "\n",
    "# Remove the repeated tissues\n",
    "low_tissue = set(low_tissue)\n",
    "medium_tissue = set(medium_tissue)\n",
    "high_tissue = set(high_tissue)\n",
    "\n",
    "# Convert the lists into strings for outputing\n",
    "low_tissue = \", \".join(low_tissue)\n",
    "medium_tissue = \", \".join(medium_tissue)\n",
    "high_tissue = \", \".join(high_tissue)    \n",
    "\n",
    "# Output the results\n",
    "print(\"Tissues with low relative variability:\", low_tissue)\n",
    "print(\"Tissues with medium relative variability:\", medium_tissue)\n",
    "print(\"Tissues with high relative variability:\", high_tissue)  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a3f6e9a",
   "metadata": {},
   "source": [
    "Answers:  \n",
    "(1) \n",
    "Tissues with low relative variability: Testis, Thyroid, Pancreas, Liver, Pituitary.  \n",
    "Tissues with medium relative variability: Prostate, Lung, Small Intestine - Terminal Ileum, Liver, Minor Salivary Gland.  \n",
    "Tissues with high relative variability: Stomach.  \n",
    "\n",
    "(2)\n",
    "Genes for a given tissue do NOT show the same level of relative variability, since Liver is on both low-relative-variability and medium-relative-variability lists. \n",
    "\n",
    "(3)\n",
    "Explanation: a low relative variability means a low stand deviation and/or a high mean. Organs like Testis, Thyroid, Pancreas, and Pituitary are important endocrine glands that produces important hormones. The tissue-specific genes in these tissues/organs are likely associated with hormone production and release. The genes envolved in hormone-related or metabolism-related (like in liver) pathways have low genetic variations since mutations are likely to cause malfunctions and even deadly consequences. \n",
    "\n",
    "The tissue-specific genes envolved in respiration (in lungs) and digestion (in intestine) have a medium genetic variation since they perform important functions but allow minor variations. \n",
    "\n",
    "The tissue-specific genes in stomach have high genetic variations because they are under lower natural-selection pressure and allow more variations. \n",
    "\n",
    "(4)\n",
    "I initially expected a high relative variation in endocrine glands like Pancreas and Pituitary, since they are not always secreting hormones. Insulin and glucagon are only secreted at high levels at certain periods of time during a day, depending on the time after the last meal. Secretion of growth hormone is high during teenage and follows circadian rhythm. All these factors may result in a high relative variation, if the samples are collected at different times during a day or on people at different ages. To my surprise, they show low relative variations, probably because the conditions of samples are well-controlled, like ages, sex, and time of collection. \n",
    "\n",
    "Besides, why stomach has such high relative variations is also surprising. I need more information on the genes to figure out the reason why they are under lower natural-selection pressure and allow more variations. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qb25",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
